app/models/user.rb

```diff
@@ -121,6 +120,9 @@ def self.name_format
 
   dirty_has_many_associations :roles
 
+  audited :except => [:last_login_on, :password_hash, :password_salt, :password_confirmation],
+          :associations => :roles
```

Comment:
here we're lucky that we already have `dirty_has_many_associations` definition but I wonder if it wouldn't be better if `audited` call should not include DirtyAssociations and add the definition for all associations. If you look at usergroup model, that would mean we don't need to add line 8 and 43. I think it would be less prone for errors in future.

Reply:
associational audits utilizes `send("#{association_ids}_changed?")` and `send("#{association_ids}_change")` provided by `dirty_has_many_associations`.

Are you suggesting to `include DirtyAssociations` in `AuditAssociations` itself? Or remove the dependency of `DirtyAssociations` and track dirtiness of associations in a different way.

Reply:
I think it would be good if DirtyAssociations were included in AuditAssociations and by calling `audited :associations => :a, :b` it would automatically define `dirty_has_many_associations :a, :b`

Reply:
Ah! right, makes sense. Thanks @ares

Reply:
This change has been done.

Reply:
I'm not sure about this tbh, we sometimes want to be able to track changes to associations, for example to check things in callbacks, not just for auditing. for example - https://github.com/theforeman/foreman/blob/develop/app/models/user.rb#L373
Audits should just audit the changes, and with dirty associations we should make sure they are set correctly for the auditing callbacks.

Reply:
I'm not sure I understand your concern, you can still use dirty cache without auditing. But if you specify association for auditing, it would automatically add dirty cache definition for you.